gh-101100: Fix Sphinx warnings in `argparse` module (#103289)
python/cpython: Doc/library/argparse.rst

.. versionchanged:: 3.12
   :class:`ArgumentParser` changed encoding and errors to read arguments files

>    from default (e.g. :func:`locale.getpreferredencoding(False) <locale.getpreferredencoding()>` and

Reviewer: ```suggestion
   from default (e.g. :func:`locale.getpreferredencoding(False) <locale.getpreferredencoding>` and
```

NB, the parens are superfluous in func/meth ref names (and the canonical form is without them).

>    from default (e.g. :func:`locale.getpreferredencoding(False) <locale.getpreferredencoding()>` and

Reviewer 2: ```suggestion
   from default (e.g. :func:`locale.getpreferredencoding(False) <locale.getpreferredencoding>` and
```
Are the `()` needed here?

Reviewer: Yeah, nope. My identical suggestion also removes them; feel free to apply either one of course.

   :class:`ArgumentParser` supports the creation of such sub-commands with the
   :meth:`add_subparsers` method.  The :meth:`add_subparsers` method is normally
   called with no arguments and returns a special action object.  This object

>    has a single method, :meth:`!add_parser`, which takes a

Reviewer: ```suggestion
   has a single method, :meth:`add_parser`, which takes a
```

Unsilence this, since it is a legitimate warning of missing documentation

Reviewer 3: *(I saw this PR on the discord and it is now the last part of my 'day of procrastination' -- sorry for my absence of late, hopefully will be more around towards the end of the month)*

As an alternate, given that nearly everything in ``argparse.rst`` is fixed, we could add an entry to ``nitpick_ignore`` in ``conf.py`` with a detailed explanation, e.g.:

```python
nitpick_ignore = [
    # do not error nit-picky mode builds when ArgumentParser.add_parser cannot
    # be resolved, as the method is currently undocumented. For context, see
    # https://github.com/python/cpython/pull/103289.
    ('py:meth', 'ArgumentParser.add_parser'),
]
```

This would require reverting the above (and the usage on line 1792 below) to `` :meth:`~ArgumentParser.add_parser` ``, and would have the same benefits that once `.add_parser()` is documented, everything would work as if by magic.

[`nitpick_ignore` reference docs](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-nitpick_ignore)

A

Reviewer: Good to see you back, and thanks for the input and ideas. It sounds like that approach could definitely make sense, particularly in cases where the existing object has the correct target (or can be easily corrected to such), but that target is just not yet documented and doing so is deferred to a followup PR. 

In this case, `ArgumentParser.add_parser` was actually a one-off typo (the other instances just feature a bare method, which this harmonizes them to), as `add_parser` is not a method of `ArgumentParser` at all, but rather a separate subparsers object (of class `_SubParsersAction`) returned by calling `ArgumentParser.add_subparsers()`. In order to implement this, we would need to fix the existing references to point to the actual target, like so:

```suggestion
   has a single method, :meth:`~_SubParsersAction.add_parser`, which takes a
```

and add it to `nitpick_ignore` as you suggest above; then, when we document it just remove it from `.nitignore`. This is slightly less trivial than just keeping the status quo for this PR and changing the refs in the followup, and does mean an implicit commitment to implementing the specific solution I propose above (documenting it as `.. method:: _SubParsersAction.add_parser`), but up to @hugovk wants to implement that approach here vs. just reverting the silencing for now; I added alternate suggestions accordingly.

   for that particular parser will be printed.  The help message will not
   include parent parser or sibling parser messages.  (A help message for each
   subparser command, however, can be given by supplying the ``help=`` argument

>    to :meth:`!add_parser` as above.)

Reviewer: ```suggestion
   to :meth:`add_parser` as above.)
```

Likewise

Reviewer: Or, if we want to implement @AA-Turner 's approach and prepare this for the followup:

```suggestion
   to :meth:`~_SubParsersAction.add_parser` as above.)
```

* Replace the OptionParser constructor ``version`` argument with a call to
  ``parser.add_argument('--version', action='version', version='<the version>')``.

Exceptions
----------

.. exception:: ArgumentError

   An error from creating or using an argument (optional or positional).

   The string value of this exception is the message, augmented with
   information about the argument that caused it.

.. exception:: ArgumentTypeError

>    An error from trying to convert a command line string to a type.

Reviewer 2: `An error from trying to ...` sounds a bit awkward to me.  `Raised if/while/when ...` like you did below sounds better.

Reviewer: What about

```suggestion
   Raised when something goes wrong converting a command line string to a type.
```

Author: Sounds good.

This was lifted from the docstring, shall we update it to match?

https://github.com/python/cpython/blob/4dc339b4d69195448207e1faecc3e258700daf33/Lib/argparse.py#L781-L783
